sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala

```diff
@@ -317,14 +318,39 @@ private[sql] class DynamicPartitionWriterContainer(
     isAppend: Boolean)
   extends BaseWriterContainer(relation, job, isAppend) {
 
+  private def toAttribute(columnName: String) = inputSchema.find(_.name == columnName).get
+
   def writeRows(taskContext: TaskContext, iterator: Iterator[InternalRow]): Unit = {
     val outputWriters = new java.util.HashMap[InternalRow, OutputWriter]
     executorSideSetup(taskContext)
 
     var outputWritersCleared = false
 
-    // Returns the partition key given an input row
-    val getPartitionKey = UnsafeProjection.create(partitionColumns, inputSchema)
+    val getKey = if (bucketSpec.isEmpty) {
+      UnsafeProjection.create(partitionColumns, inputSchema)
+    } else {
+      val BucketSpec(numBuckets, bucketColumns, sortColumns) = bucketSpec.get
+      val bucketIdExpr = Pmod(Hash(bucketColumns.map(toAttribute)), Literal(numBuckets))
+      val sortingAttrs = sortColumns.map(_.map(toAttribute)).getOrElse(Nil)
+      UnsafeProjection.create(partitionColumns ++ (bucketIdExpr +: sortingAttrs), inputSchema)
```

Comment:
The logic is: if we have sorting columns, then we will immediately sort input rows, and won't use the hash map to keep all writers.